### PR TITLE
feat(expo): add built-in TextDecoder API for native

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- This is needed for Expo Router's React Server Components support.

# How

- Pull in a subset of `text-encoding` which includes only the utf-8.
- Skip TextEncoder since this is included in Hermes, server, and web platforms already.

# Test Plan

- Added unit tests for the utf8 decoding and error classes.

### CODE REVIEW

File: packages/expo/src/winter/TextDecoder.ts [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+  }
+  return null;
+}
+
+/** [Encodings table](https://encoding.spec.whatwg.org/encodings.json) (Incomplete as we only need TextDecoder utf8 in Expo RSC. A more complete implementation should be added to Hermes as native code.) */
+const ENCODING_MAP: { heading: string; encodings: Encoding[] }[] = [
+  {
+    encodings: [
+      {
+        labels: ['unicode-1-1-utf-8', 'utf-8', 'utf8'],
```
Comment (ide) [inline packages/expo/src/winter/TextDecoder.ts:148]: These are all in the spec. Did you intentionally omit some of the values or can they be included? They all should map to the same UTF-8 encoding.
```
"unicode-1-1-utf-8"
"unicode11utf8"
"unicode20utf8"
"utf-8"
"utf8"
"x-unicode20utf8"
```

**Review by ide — APPROVED** [review]

File: packages/expo/src/winter/TextDecoder.ts (hunk 1 above)
Comment (EvanBacon) [reply packages/expo/src/winter/TextDecoder.ts:148]: Good call

**Review by tsapeta — APPROVED** [review]

File: docs/pages/versions/unversioned/sdk/encoding.mdx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+The standard URL API is available on all Expo-supported platforms.
+
+## Installation
+
+This API is part of the `expo` package. Refer to the [browser and server runtime support](https://caniuse.com/textencoder) for web and Node.js.
+
+## Usage
+
+`TextEncoder` and `TextDecoder` are available globally without needing to import them.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/encoding.mdx:17]:
```suggestion
`TextEncoder` and `TextDecoder` are available globally without any need to import them.
```

Just a small suggestion to add a bit more clarity.

File: docs/pages/versions/unversioned/sdk/encoding.mdx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+// [104, 101, 108, 108, 111]
+const hello = new TextEncoder().encode('hello');
+
+// "hello"
+const text = new TextDecoder().decode(hello);
+```
+
+## TextEncoder
+
+The `TextEncoder` API is included in the Hermes engine. See the source code here: [facebook/hermes > TextEncoder.cpp](https://github.com/facebook/hermes/blob/9e2bbf8eda15936ee00aee4f8e024ceaa7cd800d/lib/VM/JSLib/TextEncoder.cpp#L1).
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/encoding.mdx:29]:
```suggestion
The `TextEncoder` API is included in the Hermes engine. See the [source code in TextEncoder.cpp inside Hermes GitHub repository](https://github.com/facebook/hermes/blob/9e2bbf8eda15936ee00aee4f8e024ceaa7cd800d/lib/VM/JSLib/TextEncoder.cpp#L1).
```

Let's spell out where the source code is and what file it is.

**Review by amandeepmittal — APPROVED** [review]
Lgtm 👍 
Left a couple of suggestions to improve readability.

**Review by tsapeta — COMMENTED** [review]
It might be a good idea to add the documentation to our new page that @Simek is building for the `expo` package: https://github.com/expo/expo/blob/main/docs/pages/versions/unversioned/sdk/expo.mdx

See https://docs.expo.dev/versions/unversioned/sdk/expo

File: docs/pages/versions/unversioned/sdk/encoding.mdx (hunk 3 above)
Comment (ide) [reply docs/pages/versions/unversioned/sdk/encoding.mdx:29]: "inside the Hermes GitHub repository"

Comment (amandeepmittal) [reply docs/pages/versions/unversioned/sdk/encoding.mdx:29]: Thanks, update the above suggestion.

File: packages/expo/src/winter/TextDecoder.ts [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+// `TextEncoder` is in Hermes and we only need utf-8 decoder for React Server Components.
+//
+// https://github.com/inexorabletash/text-encoding/blob/3f330964c0e97e1ed344c2a3e963f4598610a7ad/lib/encoding.js#L1
+
+/**
+ * Checks if a number is within a specified range.
+ * @param {number} a The number to test.
+ * @param {number} min The minimum value in the range, inclusive.
+ * @param {number} max The maximum value in the range, inclusive.
+ * @returns {boolean} True if a is within the range.
```
Comment (Simek) [inline packages/expo/src/winter/TextDecoder.ts:11]:
```suggestion
 * @param a The number to test.
 * @param min The minimum value in the range, inclusive.
 * @param max The maximum value in the range, inclusive.
 * @returns `true` if a passed number is within the specified range.
```

Code comments nits. We can avoid setting up the types, those are automatically resolved in docs from the TypeScript definitions.

File: packages/expo/src/winter/TextDecoder.ts [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+ */
+function inRange(a: number, min: number, max: number): boolean {
+  return min <= a && a <= max;
+}
+
+/**
+ * Converts an array of code points to a string.
+ * @param {number[]} codePoints Array of code points.
+ * @returns {string} The string representation.
```
Comment (Simek) [inline packages/expo/src/winter/TextDecoder.ts:20]:
```suggestion
 * @param codePoints Array of code points.
 * @returns The string representation of given array.
```

File: packages/expo/src/winter/TextDecoder.ts [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+    this.utf8BytesNeeded = 0;
+    this.utf8BytesSeen = 0;
+
+    // 11. Return a code point whose value is code point.
+    return code_point;
+  }
+}
+
+// 8.1 Interface TextDecoder
+
```
Comment (Simek) [inline packages/expo/src/winter/TextDecoder.ts:312]:
```suggestion
// @docsMissing
```

Let's annotate missing description of exported class, or fill it up.

**Review by Simek — APPROVED** [review]

**Comment by EvanBacon** [comment]
@tsapeta the more I think about this, the more I think it's okay if this module has it's own page. Similar to the URL page https://docs.expo.dev/versions/latest/sdk/url/ but we may want to revisit in the future to have these pages clearly marked as part of the Expo standard runtime.

**Pull request merged**